### PR TITLE
fix: table icon selector js mode toggle issue

### LINKED ISSUE #8250: [Bug] When column type is Icon and icon is selected then any modifications via JS displays a code snippet when JS is turned off
## Description
Code snippet is added to Icon JS element when any character is added to the selected keyword

### Steps to reproduce the behaviour:
1. Go to any column on table widget and select column type as **Icon Button**
2. Scroll to Button Properties and choose an icon
3. Toggle JS to see the name of the button
4. Now add any character (space, alphabets, numbers or symbols) and toggle JS again and observe code appear in the field

### Important Details
- Version: Cloud
- OS: Win10
- Browser: Chrome
- Environment: Release

[LOOM DEMO](https://www.loom.com/share/cbfdb61bf7d140c19bc17ebca5631ad3)

### PR DESCRIPTION
**Pull request by aswathkk**
## Description

The issue was happening since there was no proper validation.
As part of this added a new validation type to validate the properties of table.

Fixes #8250 

Following video explains the use of `ValidationTypes.TABLE_PROPERTY`

https://user-images.githubusercontent.com/10436935/143578533-dac8f079-9799-4770-b7fe-cba49a4401ba.mov

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Manually tested

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/table-icon-selector 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.17 **(0.03)** | 37.12 **(0.08)** | 33.91 **(0.01)** | 55.66 **(0.02)**
 :red_circle: | app/client/src/utils/validation/common.ts | 55.74 **(-0.93)** | 53.13 **(-0.33)** | 33.33 **(0)** | 55 **(-0.93)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/propertyConfig.ts | 96 **(0.26)** | 100 **(0)** | 93.55 **(0.22)** | 96 **(0.26)**
 :green_circle: | app/client/src/workers/validations.ts | 77.23 **(2.61)** | 71.55 **(1.67)** | 67.65 **(2.94)** | 78.78 **(2.19)**</details>

### CODE REVIEW

**Review by riodeuno — DISMISSED** [review]

**Review by riodeuno — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@aswathkk It's working fine for icon type property but I observe issue for button colour and shadow colour properties.
https://www.loom.com/share/89301957c9a14a75acd64157b67578f1

File: app/client/src/utils/validation/common.ts [hunk 1]
```diff
@@ -156,5 +156,13 @@ export function getExpectedValue(
         example: `https://www.example.com`,
         autocompleteDataType: AutocompleteDataType.STRING,
       };
+    case ValidationTypes.TABLE_PROPERTY:
+      return getExpectedValue(config.params as ValidationConfig);
+    case ValidationTypes.COLOR:
```
Comment (riodeuno) [inline app/client/src/utils/validation/common.ts:161]: Can we use a text validation with regex for color here? The video demo for widget development uses this.

**Review by riodeuno — CHANGES_REQUESTED** [review]
One concern about the color validation. The rest looks good.

File: app/client/src/utils/validation/common.ts (hunk 1 above)
Comment (aswathkk) [reply app/client/src/utils/validation/common.ts:161]: I thought color can be any color value that we use in CSS.
Should we restrict to only hex?

File: app/client/src/entities/Widget/utils.test.ts [hunk 2]
```diff
@@ -199,6 +201,15 @@ describe("getAllPathsFromPropertyConfig", () => {
         isVisible: {
           type: "BOOLEAN",
         },
+        "primaryColumns.status.buttonColor": {
+          type: ValidationTypes.TABLE_PROPERTY,
+          params: {
+            type: ValidationTypes.COLOR,
```
Comment (riodeuno) [inline app/client/src/entities/Widget/utils.test.ts:207]: We should remove this now, or the tests should fail. Is that correct?

Comment (aswathkk) [reply app/client/src/entities/Widget/utils.test.ts:207]: my bad. Will update it now

**Review by riodeuno — DISMISSED** [review]

**Comment by vivekverma2312** [comment]
@aswathkk Thanks for the fix, it's working fine on deploy preview.

File: app/client/src/widgets/constants.ts [hunk 3]
```diff
@@ -26,3 +26,5 @@ export type AlignWidget = "LEFT" | "RIGHT";
 
 // Minimum Rows for Widget Popups
 export const MinimumPopupRows = 12;
+
+export const COLOR_REGEX = /^(#([0-9a-f]{2}){2,4}|#[0-9a-f]{3}|(rgb)\((\s*\d+\s*,){2}\s*\d+\s*\)|(rgba)\((\s*\d+\s*,){3}\s*[\d.]+\s*\))$/i;
```
Comment (riodeuno) [inline app/client/src/widgets/constants.ts:30]: We need tests for this regex. Let's add some jest tests which validate different types of colour inputs with this regex.

**Review by riodeuno — DISMISSED** [review]

File: app/client/src/widgets/constants.ts (hunk 3 above)
Comment (aswathkk) [reply app/client/src/widgets/constants.ts:30]: @riodeuno Its available at app/client/src/widgets/constants.test.ts

**Comment by riodeuno** [comment]
Ah sorry, I didn't see that

**Review by riodeuno — DISMISSED** [review]
LGTM!
Let's make sure to go through a manual regression for all table columns

**Comment by vivekverma2312** [comment]
@aswathkk The issue exist for Text colour and Cell back ground property for Plain text, URL, Number and Date types.
https://www.loom.com/share/532e4255d8d14726a2a1b6eac90ee426

**Review by riodeuno — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@aswathkk It's working fine for Icon selector.

**Pull request merged**